lib/enkf/tests/row_scaling.cpp

```diff
@@ -15,10 +15,152 @@
    See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
    for more details.
 */
+#include <stdlib.h>
+#include <chrono>
+#include <iostream>
+#include <stdexcept>
+#include <vector>
+
 #include <ert/util/test_util.hpp>
 
 #include <ert/enkf/row_scaling.hpp>
+#include <ert/res_util/matrix.hpp>
+#include <ert/util/rng.hpp>
+
+
+void test_create() {
+  row_scaling_type * row_scaling = row_scaling_alloc();
+  test_assert_int_equal(row_scaling_get_size(row_scaling), 0);
+
+  test_assert_throw(row_scaling_iget(row_scaling, -1), std::out_of_range);
+  test_assert_throw(row_scaling_iget(row_scaling, 1000), std::out_of_range);
+
+  test_assert_throw(row_scaling_iset(row_scaling, -1, 0), std::out_of_range);
+  test_assert_throw(row_scaling_iset(row_scaling, 0, -1), std::invalid_argument);
+  test_assert_throw(row_scaling_iset(row_scaling, 0, 2), std::invalid_argument);
+
+  row_scaling_iset(row_scaling, 9, 0.25);
+  test_assert_double_equal(row_scaling_iget(row_scaling, 9), 0.25);
+
+  row_scaling_free(row_scaling);
+}
+
+static void scaleX(matrix_type* X, const matrix_type * X0, double alpha) {
+  matrix_assign(X, X0);
+  matrix_scale(X, alpha);
+  for (int i=0; i < matrix_get_rows(X); i++)
+    matrix_iset(X, i, i, (1 - alpha) + matrix_iget(X,i,i));
+}
+
+void row_scaling_multiply2(const row_scaling_type * row_scaling, matrix_type * A, const matrix_type * X0) {
+  matrix_type * X = matrix_alloc(matrix_get_rows(X0), matrix_get_columns(X0));
+  for (int row=0; row < row_scaling_get_size(row_scaling); row++) {
+    double alpha = row_scaling_iget(row_scaling, row);
+    scaleX(X, X0, alpha);
+
+    std::vector<double> row_data(matrix_get_columns(A));
+    for (int j=0; j < matrix_get_columns(A); j++) {
+      double sum = 0;
+      for (int i=0; i < matrix_get_columns(A); i++)
+        sum += matrix_iget(A, row, i) * matrix_iget(X,i,j);
+
+      row_data[j] = sum;
+    }
+    matrix_set_row(A, row_data.data(), row);
+  }
+  matrix_free(X);
+}
+
+void test_multiply(const row_scaling_type * row_scaling, const matrix_type * A0, const matrix_type * X0) {
+  matrix_type * A1 = matrix_alloc_copy(A0);
+  matrix_type * A2 = matrix_alloc_copy(A0);
+
+  row_scaling_multiply(row_scaling, A1, X0);
+  row_scaling_multiply(row_scaling, A2, X0);
```

Comment:
I assume one of these should be row_scaling_multiply2?

Reply:
Ahhhhh  ;-)